File: software/host/src/storage/DataLogger.cpp
/**
 * HeteroLink Host - 数据记录器实现
 */

#include "storage/DataLogger.h"
#include "utils/Logger.h"
#include <QDateTime>
#include <QDir>
#include <QFileInfo>

namespace HeteroLink {

DataLogger::DataLogger(QObject *parent)
    : QObject(parent)
{
}

DataLogger::~DataLogger()
{
    stopRecording();
}

bool DataLogger::startRecording(const QString& basePath, const QString& deviceId)
{
    if (recording_) {
        LOG_WARNING("Already recording");
        return false;
    }
    
    basePath_ = basePath;
    deviceId_ = deviceId;
    
    // 创建目录
    QDir dir(basePath);
    if (!dir.exists()) {
        if (!dir.mkpath(".")) {
            LOG_ERROR("Failed to create directory: " + basePath.toStdString());
            return false;
        }
    }
    
    if (!openFile()) {
        return false;
    }
    
    recording_ = true;
    bytesWritten_ = 0;
    lastSplitTime_ = QDateTime::currentMSecsSinceEpoch();
    
    LOG_INFO("Recording started: " + currentFilePath().toStdString());
    emit recordingStarted(currentFilePath());
    
    return true;
}

void DataLogger::stopRecording()
{
    if (!recording_) {
        return;
    }
    
    closeFile();
    recording_ = false;
    
    LOG_INFO("Recording stopped");
    emit recordingStopped();
}

bool DataLogger::isRecording() const
{
    return recording_;
}

QString DataLogger::currentFilePath() const
{
    if (file_ && file_->isOpen()) {
        return file_->fileName();
    }
    return "";
}

void DataLogger::setMaxFileSize(int sizeMB)
{
    maxFileSize_ = sizeMB * 1024 * 1024;  // 转换为字节
}

void DataLogger::setAutoSplit(bool enabled, int intervalMs)
{
    autoSplit_ = enabled;
    splitInterval_ = intervalMs;
}

void DataLogger::writeData(const QString& deviceId, const TelemetryData& data)
{
    if (!recording_ || deviceId != deviceId_) {
        return;
    }
    
    if (!stream_) {
        return;
    }
    
    // 写入数据行
    QString line = QString::number(data.timestamp);
    for (const auto& ch : data.channels) {
        line += "," + QString::number(ch, 'f', 6);
    }
    line += "\n";
    
    QByteArray bytes = line.toUtf8();
    qint64 written = stream_->writeString(bytes);
    bytesWritten_ += written;
    
    // 检查是否需要分割文件
    if (autoSplit_) {
        qint64 now = QDateTime::currentMSecsSinceEpoch();
        
        // 检查文件大小
        if (bytesWritten_ >= maxFileSize_) {
            splitFile();
        }
        // 检查时间间隔
        else if (now - lastSplitTime_ >= splitInterval_) {
            splitFile();
        }
    }
}

bool DataLogger::openFile()
{
    QString filePath = generateFilePath();
    
    file_ = std::make_unique<QFile>(filePath);
    if (!file_->open(QIODevice::WriteOnly | QIODevice::Text)) {
        LOG_ERROR("Failed to open file: " + filePath.toStdString());
        emit errorOccurred("Failed to open file: " + file_->errorString());
        file_.reset();
        stream_.reset();
        return false;
    }
    
    stream_ = std::make_unique<QTextStream>(file_.get());
    writeHeader();
    
    return true;
}

void DataLogger::closeFile()
{
    if (stream_) {
        stream_->flush();
        stream_.reset();
    }
    
    if (file_) {
        file_->close();
        file_.reset();
    }
}

QString DataLogger::generateFilePath() const
{
    QDateTime now = QDateTime::currentDateTime();
    QString timestamp = now.toString("yyyyMMdd_HHmmss");
    QString fileName = QString("heterolink_%1_%2.csv").arg(deviceId_).arg(timestamp);
    return basePath_ + "/" + fileName;
}

void DataLogger::writeHeader()
{
    if (!stream_) {
        return;
    }
    
    // 写入 CSV 表头（假设最多 16 个通道）
    QString header = "timestamp";
    for (int i = 0; i < 16; ++i) {
        header += ",ch" + QString::number(i);
    }
    header += "\n";
    
    stream_->writeString(header.toUtf8());
    stream_->flush();
}

void DataLogger::splitFile()
{
    QString oldPath = file_->fileName();
    
    closeFile();
    
    if (!openFile()) {
        return;
    }
    
    bytesWritten_ = 0;
    lastSplitTime_ = QDateTime::currentMSecsSinceEpoch();
    
    LOG_INFO("File split: " + oldPath.toStdString() + " -> " + currentFilePath().toStdString());
    emit fileSplit(oldPath, currentFilePath());
}

} // namespace HeteroLink

File: software/host/src/storage/data_logger.rs
use crate::protocol::TelemetryData;
use anyhow::Result;
use chrono::Local;
use std::fs::{self, File};
use std::io::Write;
use std::path::PathBuf;
use std::time::{SystemTime, UNIX_EPOCH};
use tracing::{info, warn, error};

/// 数据记录器
pub struct DataLogger {
    file: Option<File>,
    base_path: String,
    device_id: String,
    recording: bool,
    max_file_size: u64,
    auto_split: bool,
    split_interval_ms: u64,
    bytes_written: u64,
    last_split_time: u64,
}

impl DataLogger {
    pub fn new() -> Self {
        Self {
            file: None,
            base_path: String::new(),
            device_id: String::new(),
            recording: false,
            max_file_size: 100 * 1024 * 1024,
            auto_split: false,
            split_interval_ms: 3600000,
            bytes_written: 0,
            last_split_time: 0,
        }
    }

    pub fn start_recording(&mut self, base_path: &str, device_id: &str) -> Result<()> {
        if self.recording {
            warn!("Already recording");
            return Ok(());
        }

        self.base_path = base_path.to_string();
        self.device_id = device_id.to_string();

        fs::create_dir_all(base_path)?;

        self.open_file()?;
        self.recording = true;
        self.bytes_written = 0;
        self.last_split_time = SystemTime::now()
            .duration_since(UNIX_EPOCH)
            .unwrap()
            .as_millis() as u64;

        info!("Recording started: {:?}", self.current_file_path());
        Ok(())
    }

    pub fn stop_recording(&mut self) {
        if !self.recording {
            return;
        }
        self.close_file();
        self.recording = false;
        info!("Recording stopped");
    }

    pub fn is_recording(&self) -> bool {
        self.recording
    }

    pub fn current_file_path(&self) -> Option<PathBuf> {
        if self.recording {
            Some(PathBuf::from(&self.base_path).join(self.generate_filename()))
        } else {
            None
        }
    }

    pub fn set_max_file_size(&mut self, size_mb: u64) {
        self.max_file_size = size_mb * 1024 * 1024;
    }

    pub fn set_auto_split(&mut self, enabled: bool, interval_ms: u64) {
        self.auto_split = enabled;
        self.split_interval_ms = interval_ms;
    }

    pub fn write_data(&mut self, device_id: &str, data: &TelemetryData) {
        if !self.recording || device_id != self.device_id {
            return;
        }

        if let Some(ref mut file) = self.file {
            let mut line = data.timestamp.to_string();
            for ch in &data.channels {
                line.push_str(&format!(",{:.6}", ch));
            }
            line.push('\n');

            let bytes = line.as_bytes();
            if let Err(e) = file.write_all(bytes) {
                error!("Failed to write data: {}", e);
                return;
            }
            self.bytes_written += bytes.len() as u64;

            if self.auto_split {
                let now = SystemTime::now()
                    .duration_since(UNIX_EPOCH)
                    .unwrap()
                    .as_millis() as u64;

                if self.bytes_written >= self.max_file_size {
                    let _ = self.split_file();
                } else if now - self.last_split_time >= self.split_interval_ms {
                    let _ = self.split_file();
                }
            }
        }
    }

    fn open_file(&mut self) -> Result<()> {
        let path = PathBuf::from(&self.base_path).join(self.generate_filename());
        let mut file = File::create(&path)?;

        // 写入 CSV 表头
        let mut header = "timestamp".to_string();
        for i in 0..16 {
            header.push_str(&format!(",ch{}", i));
        }
        header.push('\n');
        file.write_all(header.as_bytes())?;

        self.file = Some(file);
        Ok(())
    }

    fn close_file(&mut self) {
        self.file.take();
    }

    fn generate_filename(&self) -> String {
        let now = Local::now();
        format!("heterolink_{}_{}.csv", self.device_id, now.format("%Y%m%d_%H%M%S"))
    }

    fn split_file(&mut self) -> Result<()> {
        let old_path = self.current_file_path();
        self.close_file();
        self.open_file()?;
        self.bytes_written = 0;
        self.last_split_time = SystemTime::now()
            .duration_since(UNIX_EPOCH)
            .unwrap()
            .as_millis() as u64;

        if let Some(path) = old_path {
            info!("File split: {:?} -> {:?}", path, self.current_file_path());
        }
        Ok(())
    }
}
